File: api/src/services/searchService.ts
import MiniSearch from "minisearch";
import { sql } from "./db";

const miniSearchOptions = {
  fields: ["url", "title", "description", "snippet", "language"],
  storeFields: ["url", "title", "description", "snippet", "language", "nsfw", "crawledAt"],
  idField: "url",
  searchOptions: {
    boost: { url: 10, title: 5, description: 2, snippet: 1 },
    fuzzy: 0.15,
    prefix: true,
  },
};

let searchIndex = new MiniSearch(miniSearchOptions);
let indexLoaded = false;

export async function loadIndex() {
  try {
    const result = await sql`SELECT index_data FROM search_index WHERE id = 1`;

    if (result.length > 0 && result[0].index_data) {
      searchIndex = MiniSearch.loadJSON(
        result[0].index_data,
        miniSearchOptions,
      );
      indexLoaded = true;
      console.log(`[SearchService] Postgress loaded`);
    }
  } catch (error) {
    console.error(`[SearchService] Error loading index:`, error);
    indexLoaded = true;
  }
}

export async function search(
  query: string,
  page: number = 1,
  limit: number = 50,
  lang: string = "auto",
  safeSearch: string = "moderate",
) {
  if (!indexLoaded) return { results: [], total: 0 };

  const filter =
    lang && lang !== "auto"
      ? (result: any) => result.language === lang
      : undefined;

  let allResults = searchIndex
    .search(query, {
      boost: { url: 10, title: 5, description: 2, snippet: 1 },
      combineWith: "AND",
      prefix: false,
      fuzzy: false,
      filter,
    })
    .slice(0, 400);

  if (allResults.length < 10) {
    allResults = searchIndex
      .search(query, {
        boost: { url: 10, title: 5, description: 2, snippet: 1 },
        combineWith: "AND",
        prefix: (term: string) => term.length > 3,
        fuzzy: false,
        filter,
      })
      .slice(0, 400);
  }

  if (allResults.length === 0) {
    allResults = searchIndex
      .search(query, {
        boost: { url: 10, title: 5, description: 2, snippet: 1 },
        combineWith: "OR",
        prefix: true,
        fuzzy: 0.2,
        filter,
      })
      .slice(0, 100);
  }

  let filteredResults = allResults;
  if (safeSearch === "strict") {
    filteredResults = allResults.filter((r: any) => !r.nsfw);
  }

  const processedResults = filteredResults
    .map((r: any) => {
      let finalScore = r.score;
      const lowerQuery = query.toLowerCase();
      const lowerTitle = (r.title || "").toLowerCase();
      const urlString = r.url.toLowerCase();

      if (urlString.includes(lowerQuery)) finalScore *= 5;
      if (lowerTitle === lowerQuery) finalScore *= 2;
      if (urlString.length > 100) finalScore *= 0.8;

      r.score = finalScore;

      if (safeSearch === "moderate" && r.nsfw) {
        r.blur = true;
      }

      return r;
    })
    .sort((a: any, b: any) => b.score - a.score);

  const groupedResults: any[] = [];
  const domainMap = new Map<string, any>();

  for (const r of processedResults) {
    try {
      const url = new URL(r.url);
      const domain = url.hostname;

      if (!domainMap.has(domain)) {
        r.sitelinks = [];
        groupedResults.push(r);
        domainMap.set(domain, r);
      } else {
        const parent = domainMap.get(domain);
        if (parent.sitelinks.length < 4) {
          parent.sitelinks.push(r);
        }
      }
    } catch (e) {
      groupedResults.push(r);
    }
  }

  const total = groupedResults.length;
  const offset = (page - 1) * limit;

  let correction: string | undefined;
  if (total < 500 && query.length >= 3) {
    let suggestions = searchIndex.autoSuggest(query, { fuzzy: 0 });

    if (suggestions.length === 0 && total < 20) {
      suggestions = searchIndex.autoSuggest(query, { fuzzy: 0.2 });
    }

    const bestSuggest = suggestions.find(
      (s: any) => s.suggestion.toLowerCase() !== query.toLowerCase(),
    );
    if (bestSuggest) {
      correction = bestSuggest.suggestion;
    }
  }

  return {
    total,
    results: groupedResults.slice(offset, offset + limit),
    totalPages: Math.ceil(total / limit),
    correction,
  };
}

export function autocomplete(query: string, limit: number = 8) {
  if (!indexLoaded || !query || query.length < 2) return [];

  const suggestions = searchIndex.autoSuggest(query, {
    fuzzy: (term: string) => (term.length > 5 ? 0.2 : 0),
    prefix: true,
    combineWith: "AND",
  });

  return suggestions.slice(0, limit).map((s: any) => s.suggestion);
}

export async function getStats() {
  try {
    const pagesCount = await sql`SELECT COUNT(*)::int AS count FROM pages`;
    const indexInfo =
      await sql`SELECT id, LENGTH(index_data)::int as size FROM search_index WHERE id = 1`;

    return {
      total_pages: pagesCount[0]?.count || 0,
      index_ready: indexLoaded,
      index_size: indexInfo[0]?.size || 0,
      total_indexed: searchIndex.documentCount,
    };
  } catch (err) {
    return null;
  }
}

File: services/searchService.go
package services

import (
	"context"
	"fmt"
	"net/url"
	"sort"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
)

type SearchDocument struct {
	URL         string            `json:"url"`
	Title       string            `json:"title"`
	Description string            `json:"description"`
	Snippet     string            `json:"snippet"`
	Language    string            `json:"language"`
	NSFW        bool              `json:"nsfw"`
	Score       float64           `json:"score"`
	Blur        bool              `json:"blur"`
	Sitelinks   []*SearchDocument `json:"sitelinks"`
}

type SearchService struct {
	db *pgx.Conn
}

func NewSearchService(dbConn *pgx.Conn) *SearchService {
	return &SearchService{db: dbConn}
}

func (s *SearchService) Search(query string, page, limit int, safeSearch string) (map[string]interface{}, error) {
	startTime := time.Now()

	if strings.TrimSpace(query) == "" {
		return map[string]interface{}{"results": []interface{}{}, "total": 0, "timeMs": 0}, nil
	}

	offset := (page - 1) * limit

	rows, err := s.db.Query(context.Background(), `
		SELECT
			url, title, description, snippet, language, nsfw,
			ts_rank(search_vector, plainto_tsquery('simple', $1)) AS rank
		FROM pages
		WHERE search_vector @@ plainto_tsquery('simple', $1)
		ORDER BY
			CASE WHEN url ILIKE '%' || $1 || '%' THEN 2 ELSE 0 END +
			CASE WHEN title ILIKE '%' || $1 || '%' THEN 1 ELSE 0 END +
			ts_rank(search_vector, plainto_tsquery('simple', $1)) DESC
		LIMIT $2 OFFSET $3
	`, query, limit*20, offset)
	if err != nil {
		return nil, fmt.Errorf("search query failed: %v", err)
	}
	defer rows.Close()

	var all []*SearchDocument
	for rows.Next() {
		var doc SearchDocument
		var rank float64
		err := rows.Scan(
			&doc.URL, &doc.Title, &doc.Description, &doc.Snippet,
			&doc.Language, &doc.NSFW, &rank,
		)
		if err != nil {
			continue
		}
		doc.Score = rank

		if safeSearch == "strict" && doc.NSFW {
			continue
		}
		if safeSearch == "moderate" && doc.NSFW {
			doc.Blur = true
		}
		all = append(all, &doc)
	}

	sort.Slice(all, func(i, j int) bool { return all[i].Score > all[j].Score })

	var grouped []*SearchDocument
	domainMap := make(map[string]*SearchDocument)

	for _, r := range all {
		u, _ := url.Parse(r.URL)
		domain := u.Hostname()
		if parent, exists := domainMap[domain]; !exists {
			r.Sitelinks = []*SearchDocument{}
			grouped = append(grouped, r)
			domainMap[domain] = r
		} else if len(parent.Sitelinks) < 4 {
			parent.Sitelinks = append(parent.Sitelinks, r)
		}
	}

	total := len(grouped)
	start := 0
	end := limit
	if end > total {
		end = total
	}

	durationMs := time.Since(startTime).Milliseconds()

	return map[string]interface{}{
		"total":       total,
		"total_pages": (total + limit - 1) / limit,
		"timeMs":      durationMs,
		"results":     grouped[start:end],
	}, nil
}

func (s *SearchService) Autocomplete(query string, limit int) []string {
	if len(strings.TrimSpace(query)) < 2 {
		return []string{}
	}

	rows, err := s.db.Query(context.Background(), `
		SELECT DISTINCT title
		FROM pages
		WHERE title ILIKE $1 || '%'
		ORDER BY title
		LIMIT $2
	`, query, limit)
	if err != nil {
		return []string{}
	}
	defer rows.Close()

	var suggestions []string
	for rows.Next() {
		var title string
		if err := rows.Scan(&title); err == nil && title != "" {
			suggestions = append(suggestions, title)
		}
	}
	return suggestions
}

func (s *SearchService) GetStats(ctx context.Context) map[string]interface{} {
	var pagesCount int
	var indexedCount int

	s.db.QueryRow(ctx, "SELECT COUNT(*)::int FROM pages").Scan(&pagesCount)
	s.db.QueryRow(ctx, "SELECT COUNT(*)::int FROM pages WHERE search_vector IS NOT NULL").Scan(&indexedCount)

	return map[string]interface{}{
		"total_pages":   pagesCount,
		"total_indexed": indexedCount,
		"index_ready":   indexedCount > 0,
	}
}
